### PR TITLE
Change per-session limits for spans started to 500 overall

### PR DESCRIPTION
**Pull request by bidetofevil**
## Goal

Only check the total number of spans in a session when apply limits

### CODE REVIEW

**Comment by bidetofevil** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/403?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `master`
  * **PR #391** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/391?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #392** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/392?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #393** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/393?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #394** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/394?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #395** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/395?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #396** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/396?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #397** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/397?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                * **PR #401** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/401?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                  * **PR #403** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/403?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
                  * **PR #402** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/402?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/403?utm_source=stack-comment).

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/CurrentSessionSpanImpl.kt [hunk 1]
```diff
@@ -55,52 +42,23 @@ internal class CurrentSessionSpanImpl(
...
      */
     override fun canStartNewSpan(parent: EmbraceSpan?, internal: Boolean): Boolean {
+        // Check conditions where no span can be created
         val currentSession = sessionSpan.get() ?: return false
         if (!currentSession.isRecording || (parent != null && parent.spanId == null)) {
             return false
         }
 
-        if (internal) {
+        // If a span can be created, always let internal spans be to be created
```
Comment (fractalwrench) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/CurrentSessionSpanImpl.kt:51]: It seems sensible to add some very high limit (e.g. 10_000)

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpansServiceImpl.kt [hunk 2]
```diff
@@ -98,7 +98,6 @@ internal class SpansServiceImpl(
     override fun getSpan(spanId: String): EmbraceSpan? = spansRepository.getSpan(spanId = spanId)
 
     companion object {
-        const val MAX_TRACE_COUNT_PER_SESSION = 100
-        const val MAX_SPAN_COUNT_PER_TRACE = 10
+        const val MAX_NON_INTERNAL_SPANS_PER_SESSION = 500
```
Comment (fractalwrench) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpansServiceImpl.kt:101]: We should tell iOS about this limit. Also, do we need similar limits in future for SpanEvent & attributes?

**Review by fractalwrench — APPROVED** [review]
LGTM

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/CurrentSessionSpanImpl.kt (hunk 1 above)
Comment (bidetofevil) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/CurrentSessionSpanImpl.kt:51]: We should. I would be an SDK error if we go beyond that. Let me add a task to track this.

The good thing is that spans are pretty small if you don't put lots of events and attributes on them. 10K internal spans is probably less in size than 100 thick spans with lots of long attributes and events. I'll do some profiling and pick a reasonable value.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpansServiceImpl.kt (hunk 2 above)
Comment (bidetofevil) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpansServiceImpl.kt:101]: iOS should be aware of the new limits and have a task to implement that in the new version.

We also have existing limits in `EmbraceSpanEvent` and `EmbraceSpanImpl`. We may want to consolidate them in a common location for clarity. I can do that in another PR.

**Comment by bidetofevil** [comment]
### Merge activity

* **Feb 14, 2:46 PM EST**: @bidetofevil started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/403).

**Pull request merged**